### PR TITLE
Vulkan: the KHR coopmat kq tile rebuilt (word stage, f16 accumulators, 2x4 tiling), emission-time [unroll_full], the recurrent f16 feed decided per activation buffer

### PR DESCRIPTION
**Pull request by borisbat**
**Behavior change: on a Vulkan device with KHR cooperative matrix and no NV_coopmat2 (`DASLLAMA_COOPMAT=mm`), K-quant prefill runs on a rebuilt tile, and a hybrid model's recurrent qkv/z GEMMs stay on it when its out plane is Q8_0.**

**Why.** Our K-quant prefill tile on the KHR arm read 32.6 TFLOP/s on the 4B gate shape against llama.cpp's 43-46, and the board's KHR prefill rows sat at 0.59-0.85x of llama.cpp; a GPU profile put the gap in the load-store pipe and the register file, not the arithmetic.

**What changes.**
- The SPIR-V emitter unrolls `for [unroll_full]` over literal bounds at emission and lays a fixed array of cooperative-matrix tiles out as one Function-storage array reached through constant chains.
- Every kq format stages its 16-value run from one or two words of the quant plane (`khr_stage16`); the tile holds f16 accumulators, eight subgroups tiled two by four.
- The recurrent head decides its f16 feed per activation buffer, not once per layer.
- The tile probe's `khrx` sweep judges itself against the shipped class and the CPU oracle, the kernel census panics on a name nothing seeded, and the review round's document fixes ride along.

**Observable behavior.**
- KHR kq tile, 4B gate / down / q shapes: 32.6 / 31.7 / 32.5 TFLOP/s -> 54.2 / 59.5 / 57.6.
- Board KHR prefill, every K-quant row now above llama.cpp: 4B Q4_K_M 3051 -> 4764 t/s, 27B UD-IQ4_XS 395 -> 741, 9B UD-Q5_K_XL 767 -> 2531; decode unchanged.
- A `for [unroll_full]` with a runtime bound, or a `break`, `continue`, `return`, `while` or second loop hint under it -> compile error.

**Where to look.** `unroll_at_emission` in `modules/dasSpirv/spirv/spirv_emit.das`; `khr_stage16` and `khr_tile` in `dasllama_vulkan_classes.das`; `pf_dnx6` / `pf_dno6` in `dasllama_vulkan_prefill.das`.

<details>
<summary>Validation, claims, ledger</summary>

### Validation
- Local-only, on the RTX 5060 Ti (driver 616.56) against llama.cpp b10660's Vulkan build the same day.
- Kernel cells (`test_vulkan_kernels.das`): 108 of 108; the thirteen per-format KHR arms each print `<fmt> khr-tile batch: 0 of 89600 off the oracle, four-wide twin` (k4, k5, q40, k2, iq4nl, iq4xs, k3, iq3s, iq2s, iq2xs, iq2xxs, iq3xxs, k6).
- The SPIR-V of every other kernel is byte-identical before and after: a `DASLLAMA_VK_SPV_DUMP` of master and of this branch differ in the thirteen `kq_batch_<fmt>_khr_spv` kernels and `dn_conv_cls_spv` only.
- Hybrid parity (`test_gpu_resident_hybrid.das`, nine cells - four on the Q8 carrier, two on the K-quant twin, two on the mixed twin, one model-free census cell; dastest counts 15): 15 of 15 under `DASLLAMA_COOPMAT=mm`, 15 of 15 on the cm2 default, 15 of 15 under `sdot4`. Every step's argmax and logits maxdiff sit inside the bar in the logs. The routing witness on the mixed twin reads `1 Q8_K requants over 18 recurrent layers, f16 feed admits kq true` on both cells under mm, and `55` to `145 Q8_K requants over 18 recurrent layers, f16 feed admits kq false` under sdot4.
- The mixed twin `Qwen3.5-0.8B-Q4_K_M-q8out.gguf` was minted by its recipe row (`llama-quantize --allow-requantize --tensor-type ssm_out=q8_0 Qwen3.5-0.8B-Q8_0.gguf Qwen3.5-0.8B-Q4_K_M-q8out.gguf Q4_K_M`), 548172032 bytes; `fetch_models.das -- -o q8out` reports `1 ok, 0 fetched, 0 pending, 0 failed` (exit 0), and the unscoped run lists the row as present (its exit 1 is other, unstocked entries).
- The K-quant parity bar on the quant feed (`sdot4`) is 10%, the f16 feed's 6%: the mixed twin's two-window cell reads 0.084 of an 11.6 max logit at one decode step there, argmax unchanged.
- Tile probe (`harness/vk_gemm_probe.das -- khrx`, exit 0): the ship, dec4, strip and wg128 copies read `bit-exact vs khr` on all three shapes; f32a reads 0.00514 / 0.00923 / 0.00525 relative, within the 2e-2 bound; the shipped class holds `0 of 768 sampled outputs off the cell bar` against the k4 CPU oracle on every shape, the partial-window row included. `khrprof:bogus` exits 1 naming the eight arms.
- Emitter suite (`tests/spirv`, 121 files): 402 of 402, the census cell with `ufor`, `ufold` and `coopmat_arr` in its roster; `test_loops.das` 15 of 15 and `test_coopmat.das` 6 of 6 with spirv-val from the local SDK (C:/VulkanSDK/1.4.357.0) - the suite's `spirv-val not found locally; skipping` line is absent from every log.
- Board rows 14 and 22 are re-runs at this tip: `lcpp_bench --for-debug-purposes -r 5 -p 512 -n 128` under `DASLLAMA_COOPMAT=mm` on 2026-09-08 (row 14 at 3a8465e2f before the rebase, the same sources); `performance/records/*.json` carries no UD-Q5_K_XL or Coder-30B row, so no record row was minted.
- Per-PR suites at the tip 7a96ebae1, through the preflight gates with the worktree binary: `--only dasllama-model-free` `[PASS] dasllama-model-free (1468.7s)`, `--only dasllama-stocked` `[PASS] dasllama-stocked` (no `--exclude`; the gate's duration print wrapped negative on a run past 35 minutes); the gate keeps the pass/fail verdict, not the runner's `REPORT` text. The full preflight ran once at 7a96ebae1: untracked, lint (64 files, both rails), hash-refs, review-md, md-ascii, ast-verify (29 files), ci-das (26 files) and compile-sweep (759 roots) PASS; format FAIL on one whitespace difference in `harness/vk_gemm_probe.das`, fixed by the formatter in the follow-up commit a62e1f3db and re-run green (`[PASS] format`). The lanes that red skipped ran one by one at a62e1f3db: docs PASS (sphinx-html 498 s), tests-cpp PASS, tests-interp 14285 of 14303 and tests-jit 14155 of 14171 with the same single failure in each - `tests/watchdog/test_watchdog.das` took 163 s against preflight's 60 s per-file cap under 48 workers, a file this diff does not touch (run alone it passes, 46 of 49 with 0 failures, in 165 s - the cap is what fails it, the ledgered per-file-allowance item); tests-aot 13226 of 13255 after building `test_aot` from a plain shell, the eleven failures all in `tests/module_cache/test_descriptor_manifest.das` - a file this branch does not touch, changed on master the same day by the commits this branch rebased onto; it passes 13 of 13 under the interpreter and fails only under the static `test_aot.exe -use-aot` runner, where the children it spawns print nothing (`marker 'PROBE 42' MISSING`, rc 0).
- The `.md` stop's rulings batch (the last commit) and the rebase onto master 5a9c46de1 ran the fast tier minus the compile sweep: untracked, format, lint, hash-refs, review-md, md-ascii, ast-verify and ci-das PASS, three gates with nothing in reach; after the dragon fixes, lint / review-md / format / md-ascii / hash-refs PASS again, and the make-pr sync, stamp-reach and jit-smoke gates are green on the rebased tip. The full preflight was not re-run after the rebase; CI validates the tip.
- CI on the tip: one red lane, `build (windows, 32, Release)`, the interpreter sweep failing only master's `tests/language/func_addr.das` (the function-address test #3974 merged the same day): the reinterpreted handle carries garbage in its high 32 bits on 32-bit, a path this diff does not touch (no change under `src/`, `include/` or the interpreter), and #3974's own 32-bit lane passed the same test, so the read is uninitialized memory. Filed as #3979; the lane was re-run once.
- The khrx arms' binding order (0 wq / 1 ws / 2 sched / 3 xf16 / 5 y) and the `BatchArgs` push-constant layout were verified by hand against `KhrPxBase` and `kq_batch_k4_khr_cls`.
- The GPU was lost once during the board's KHR re-measurement (no TDR event, rebooted, cause unknown; another session was testing the watchdog and the server on the same box); the rows that followed were re-run after the reboot. The probe's first partial-window run lost the device once more: the bisect arms store whole tiles and read past a 300-token fixture, which is why they now sit out the partial row.

### Claims - stated, not tested
- The tile is measured on one NVIDIA card in KHR mode; a wave64 device (AMD) and a KHR-only card (Turing, Intel) run the same kernels but their rates are unmeasured - the real-hardware pass is ledgered under item 42.
- The feed split is the rule the attention and FFN heads already apply, not a kernel race: one per-role profile shows both forms on one model under one instrument (qkv/z on the quant form at 11 TFLOP/s beside gate/up/down on the f16 form at 44).
- Per-format word indexing in `khr_stage16` is proven by the thirteen cells; the edge-tile store by the cells' edge tiles and the probe's partial row; f16 accumulation over the full k by the cell bars, the model bars and the cool-scales fixture (llama.cpp's own accumulator choice). The x6=false / o6=true combination (a K-quant out plane beside Q8_0 qkv/z) has no fixture; the two decisions are independent `if`s and each side is exercised.
- The emitter's fold guards (division by zero, a shift of 32 or more, INT_MIN / -1) are stated, not fixtured: the fixtures exercise the fold arms, not the refusals.
- The routing witness's threshold assumes the attention and FFN heads contribute no Q8_K requants on the fixture (they ride the f16 feed there).

### Not done
- The external codex review (the woodpecker round) did not run: codex 0.153.3's execution runner fails to start a process on this box (`timed out after 15000ms connecting runner pipe-in`) in both the `--base origin/master` and the `--commit` form, on a quiet machine as under agent load, and its fallback review tool is blocked by the never-approve policy. No external round backs this PR.
- Ledger item 42 is corrected in this change (on a KHR device llama.cpp runs its f16 coopmat tile, not the integer MMQ tile); the no-coopmat arm (`DASLLAMA_COOPMAT=sdot4`) against llama.cpp's coopmat-less arm, the wave64 twin of the KHR tile, and the real-hardware pass stay open under it.
- The decode levers on the ledger (the attention floor, the fused residual add, the fence poll) and the qwen2moe items are not this arc's.
- The `tests/spirv` census roster misses about 25 `*_words` fixtures, and the suite's skip line claims a CI check no lane makes in about a hundred files; the two files this change touches carry the corrected line, the rest and the roster go to the follow-up PR (the shader-emitter follow-up ledger, item 3). The roster rule itself lands here.
- The MoE architectures that fit a 16 GB card and need a new arch are ledgered as followup 122, ruled for after the 0.6.4 release and after the Vulkan tier covers every carrier the module already serves (followup_vulkan 43).

</details>

### CODE REVIEW

File: skills/internal/tests_in_repo.md [hunk 1]
```diff
@@ -116,3 +116,23 @@ should keep the fixture in its own directory rather than reaching across the tre
...
 Sweep the whole `tests` root for AOT/JIT validation, per the root-path caveat above.
+
+## The dasSpirv emitter suite (`tests/spirv`)
+
+`tests/spirv` is the SPIR-V emitter's own gate: every fixture compiles a shader and asserts on
+the emitted words, `test_census.das` holds the emitted opcode set equal to the declared set in
+both directions, and every blob is validated with `spirv-val` where the box resolves one - a cell
+prints `spirv-val not found locally; skipping` otherwise, and no CI lane resolves it. The run
+`modules/dasSpirv/REVIEW.md` owes, on a box with the Vulkan SDK on `PATH` (the PR body names
+it):
```
[the remark this answers is not in the document]
Comment (borisbat) [reply skills/internal/tests_in_repo.md]: Dropped the parenthetical: the sentence now states the requirement alone (the Vulkan SDK on PATH). The duty to name the run in a PR is the dasSpirv checklist's, not the skill's.

File: tests/spirv/test_coopmat.das [hunk 2]
```diff
@@ -100,4 +100,23 @@ def test_coopmat(t : T?) { // nolint:STYLE038 - flat list of opcode/capability a
...
+        t |> equal(count_op(words, SpvOp.CooperativeMatrixStoreKHR), 4, "four stores")
+        let fc = count_function_storage_chains(words)
+        t |> equal(fc.chains, 12, "twelve element chains into the tile array (read, write and convert per copy)")
+        t |> equal(fc.dynamic_indices, 0, "every chain into the tile array indexes with an OpConstant")
+        t |> success(length(words) >= 2 && words[1] == SPV_VERSION_1_6, "SPIR-V header version 1.6")
+        let r = validate_spirv(words, "vulkan1.3")
+        if (r.ran) {
+            t |> success(r.ok, "spirv-val (vulkan1.3): {r.msg}")
+        } else {
+            feint("spirv-val not found locally; skipping (CI enforces)\n")
```
[the remark this answers is not in the document]
Comment (borisbat) [reply tests/spirv/test_coopmat.das]: Fixed in test_coopmat.das and test_loops.das, the two suite files this PR touches: the skip line now reads "spirv-val not found locally; skipping - run on a box with the Vulkan SDK on PATH". The same line sits in about a hundred other files under tests/spirv, pre-existing; those go with the census-roster follow-up already ledgered.
